src/function/scalar/map/map_extract.cpp

```diff
@@ -15,11 +15,23 @@ void FillResult(Value &values, Vector &result, idx_t row) {
 	}
 
 	//! now set the pointer
-	auto &entry = ((list_entry_t *)result.GetData())[row];
+	auto &entry = ListVector::GetData(result)[row];
 	entry.length = list_values.size();
 	entry.offset = current_offset;
 }
 
+static bool ShouldCastKey(const LogicalType &map_key_type, const LogicalType &indexing_type) {
+	return map_key_type != LogicalTypeId::SQLNULL && map_key_type != indexing_type;
+}
+
+static Value GetKeyToLookFor(Vector &index_vector, idx_t row, const LogicalType &cast_to) {
+	auto key_value = index_vector.GetValue(row);
+	if (cast_to.id() != LogicalTypeId::SQLNULL) {
+		key_value = key_value.CastAs(cast_to);
```

Comment:
Rather than casting here in the function itself, we can set the required target type in the bind (`bound_function.arguments[1] = MapType::KeyType(arguments[0]->return_type)`). That way the system will insert a (more efficient and vectorized) cast and the code here can be simplified.